pkg/solana/client/client.go

```diff
@@ -0,0 +1,147 @@
+package client
+
+import (
+	"context"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/config"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/logger"
+	"golang.org/x/sync/singleflight"
+)
+
+//go:generate mockery --name ReaderWriter --output ./mocks/
+type ReaderWriter interface {
+	Writer
+	Reader
+}
+
+type Reader interface {
+	AccountReader
+	Balance(addr solana.PublicKey) (uint64, error)
+	SlotHeight() (uint64, error)
+	LatestBlockhash() (*rpc.GetLatestBlockhashResult, error)
+	ChainID() (string, error)
+	GetFeeForMessage(msg string) (uint64, error)
+}
+
+// AccountReader is an interface that allows users to pass either the solana rpc client or the relay client
+type AccountReader interface {
+	GetAccountInfoWithOpts(ctx context.Context, addr solana.PublicKey, opts *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error)
+}
+
+type Writer interface {
+	SendTx(ctx context.Context, tx *solana.Transaction) (solana.Signature, error)
+}
+
+var _ ReaderWriter = (*Client)(nil)
+
+type Client struct {
+	rpc             *rpc.Client
+	skipPreflight   bool // to enable or disable preflight checks
+	commitment      rpc.CommitmentType
+	txTimeout       time.Duration
+	contextDuration time.Duration
+	log             logger.Logger
+
+	// provides a duplicate function call suppression mechanism
+	requestGroup *singleflight.Group
+}
+
+func NewClient(endpoint string, cfg config.Config, requestTimeout time.Duration, log logger.Logger) (*Client, error) {
+	return &Client{
+		rpc:             rpc.New(endpoint),
+		skipPreflight:   cfg.SkipPreflight(),
+		commitment:      cfg.Commitment(),
+		txTimeout:       cfg.TxTimeout(),
+		contextDuration: requestTimeout,
+		log:             log,
+		requestGroup:    &singleflight.Group{},
+	}, nil
+}
+
+func (c *Client) Balance(addr solana.PublicKey) (uint64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.contextDuration)
+	defer cancel()
+
+	v, err, _ := c.requestGroup.Do("GetBalance", func() (interface{}, error) {
```

Comment:
Coalescing here depends on the input `addr` parm, so having a general `"GetBalance"` group will group them all, ignoring the arg. 

We should include the arg in the group name as `"GetBalance(${addr})"`.

Reply:
ooooh good point, otherwise it would return incorrectly if you were querying for two different balances

Reply:
fixed: https://github.com/smartcontractkit/chainlink-solana/pull/250/commits/b0b22cdca27331bc30a484ff80a1f706c46ae61e